### DISCUSSION TITLE
EC_POINT_is_on_curve failure

### DISCUSSION THREAD

**Discussion by venkatesh6911** [openssl/openssl#25279] [Q&A]
Under what circumstances or configuration, does EC_POINT_is_on_curve() fail ? 
The failure trace is given below:
0  BN_ucmp (a=0xb91035f3ee0, b=0xb91035f3ec8) at crypto/bn/bn_lib.c:732
#1  0x000000000166c0d7 in ossl_ec_GFp_simple_is_on_curve (group=0xb91013f0d70, point=0xb9100b7aa50, ctx=0xb9338da8e70) at crypto/ec/ecp_smpl.c:1051
#2  0x000000000165cf27 in EC_POINT_is_on_curve (group=0xb91013f0d70, point=0xb9100b7aa50, ctx=0xb9338da8e70) at crypto/ec/ec_lib.c:1009
#3  0x000000000165cdf7 in EC_POINT_set_affine_coordinates (group=0xb91013f0d70, point=0xb9100b7aa50, x=0xb91035f3e68, y=0xb91035f3e80, ctx=0xb9338da8e70)
    at crypto/ec/ec_lib.c:874

Which corresponds to:
    i = a->top - b->top;
    if (i != 0)
        return i;

and the value of i:
(gdb) p i
$46 = -6

(gdb) p *a
$52 = {d = 0xb9338da8db0, top = 0, dmax = 7, neg = 0, flags = 0}
(gdb) p *b
$53 = {d = 0xb9101008410, top = 6, dmax = 12, neg = 0, flags = 0}

What are the possible root causes of this failure ?

I have had a customer reporting this failure with our own custom provider module when tested for handshake. The failure is seen with P-256, P-384 and P-521 curves.

**Comment by mattcaswell** [comment]
> What are the possible root causes of this failure ?

Typically invalid or corrupt x/y co-ords or a mismatch between the point and the group (e.g. the point is for another curve)

> top = 0

This is highly suspect. It suggests that the "a" value (i.e. first param) in the `BN_ucmp` operation is 0.

Backing up to the `ossl_ec_GFp_simple_is_on_curve` function we see this:

https://github.com/openssl/openssl/blob/f7fd43402c73dcc2d8fced1fca60f7da9a3a2a45/crypto/ec/ecp_smpl.c#L1047-L1051

So the square of the Y co-ordinate is ending up as zero which sounds wrong.
